libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -1864,6 +1878,109 @@ bool AP_Mission::jump_to_abort_landing_sequence(void)
+    // is a distance travelled minimization problem.  Look forward in 
+    // mission to evaluate the shortest remaining distance to land.
+
+    // Go through the mission for the nearest DO_LAND_START first as this is the most probable route 
+    // to a landing with the minimum number of WP.
+    uint16_t do_land_start_index = get_landing_sequence_start();
+    if (do_land_start_index == 0){
+        // Then no DO_LAND_START commands are in mission and normal failsafe behaviour should be maintained
+        return false;
+    }
+
+    // get our current location
+    Location current_loc;
+    if (!AP::ahrs().get_position(current_loc)) {
+        // we don't know where we are!!
+        return false;
+    }
+
+    // Get distance to landing if travelled to nearest DO_LAND_START via RTL
+    float dist_via_do_land;
+    if (!distance_to_landing(do_land_start_index, dist_via_do_land, current_loc)){
+        // cant get a valid distance to landing
+        return false;
+    }
+
+    // Get distance to landing if continue along current mission path
+    float dist_continue_to_land;
+    if (!distance_to_landing(_nav_cmd.index, dist_continue_to_land, current_loc)) {
+        // cant get a valid distance to landing
+        return false;
+    }
+
+    // Compare distances
+    if (dist_via_do_land >= dist_continue_to_land){
+        // Then the mission should carry on uninterrupted as that is the shorter distance
+        gcs().send_text(MAV_SEVERITY_NOTICE, "Rejecting RTL: closer land if mis continued");
+        return true;
+    } else {
+        // allow failsafes to interrupt the current mission
+        return false;
+    }
+}
+
+// Approximate the distance travelled to get to a landing.  DO_JUMP commands are observed in look forward.
+bool AP_Mission::distance_to_landing(uint16_t index, float &tot_distance,Location prev_loc)
+{
+    Mission_Command temp_cmd;
+    tot_distance = 0.0f;
+
+    // reset simulated jump tracking
+    for(uint8_t i=0; i<AP_MISSION_MAX_NUM_DO_JUMP_COMMANDS; i++) {
+        _jump_tracking[i].sim_num_times_run = _jump_tracking[i].num_times_run;
+    }
+
+    // run through remainder of mission to approximate a distance to landing
+    for (uint8_t i=0; i<255; i++) {
+        // search until the end of the mission command list
+        for (uint16_t cmd_index = index; cmd_index < (unsigned)_cmd_total; cmd_index++) {
+            // get next command
+            if (!get_next_cmd(cmd_index, temp_cmd, false, true)) {
+                // we got to the end of the mission
+                return false;
+            }
+            if (temp_cmd.id == MAV_CMD_DO_PARACHUTE) {
+                // parachute is a valid way to land
+                return true;
+            }
+            if (temp_cmd.id == MAV_CMD_NAV_WAYPOINT || temp_cmd.id == MAV_CMD_NAV_LAND || temp_cmd.id == MAV_CMD_NAV_VTOL_LAND || temp_cmd.id == MAV_CMD_NAV_SPLINE_WAYPOINT) {
```

Comment:
Why do you only accept these? Loiter/Loiter to alt/Loiter time are frequently used in landing sequences on plane.

Reply:
The thought was that any sort of loiter invalidates the flying distance, i guess we could do some thing with the loiter radius

Reply:
The distance calculation is now on a MIS_OPTIONS bit.  So it represents an improvement on the current behaviour.  I would like to account for loiters in the future, but would prefer for that to be a future improvement in a separate PR.